syntax = "proto3";

import "gogo.proto";

package trade.service.v1;

option go_package = "v1";

// StreamEngineService grpc
service StreamEngineService {
  // 获取单品种行情
  rpc SubscribeOneQuote(SubscribeOneQuoteReq) returns (stream MultiQuoteData);

  // 订阅聚合行情
  rpc MultiSubscribeQuote(MultiSubscribeQuoteReq) returns (stream MultiMarketStreamData);

  // demo
  rpc Demo(DemoReq) returns (stream DemoResp);

  // 设置参数
  rpc SetParams(SetParamsReq) returns (SetParamsResp);

  // 获取参数
  rpc GetParams(GetParamsReq) returns (GetParamsResp);
}

// demo
message DemoReq {
    int32 req = 1;
}

message DemoResp {
    int32 resp = 1;
}

// 设置参数
message SetParamsReq {
    int32 depth = 1;            // 聚合行情深度
    int32 frequency = 2;        // 聚合行情频率
    int32 precise = 3;          // 设置某个品种的深度（和symbol一起使用）
    string symbol = 4;          // 设置某个品种的深度（和symbol一起使用）
    int32 raw_frequency = 5;    // 发布原始品种行情频率
}

message SetParamsResp {}

// 获取参数
message GetParamsReq {
}

message GetParamsResp {
    repeated string symbols = 1;
    repeated string exchanges = 2;
    int32 depth = 3;            // 聚合行情深度
    int32 frequency = 4;        // 聚合行情频率
    int32 raw_frequency = 5;    // 发布原始品种行情频率
}

// 订阅聚合行情
message MultiSubscribeQuoteReq {}

message MultiMarketStreamData {
  repeated MarketStreamData quotes = 1 [(gogoproto.customname) = "Quotes", (gogoproto.jsontag) = "quotes"];
}

message MarketStreamData {
  string symbol = 1 [(gogoproto.customname) = "Symbol", (gogoproto.jsontag) = "symbol"];
  string msg_seq = 2 [(gogoproto.customname) = "MsgSeq", (gogoproto.jsontag) = "msg_seq"];
  int64 time = 3 [(gogoproto.customname) = "Time", (gogoproto.jsontag) = "time", (gogoproto.casttype) = "wx/pkg/time.Time"];
  int64 time_arrive = 4 [(gogoproto.customname) = "TimeArrive", (gogoproto.jsontag) = "time_arrive", (gogoproto.casttype) = "wx/pkg/time.Time"];
  repeated Depth ask_depths = 5 [(gogoproto.customname) = "AskDepths", (gogoproto.jsontag) = "ask_depths"];
  repeated Depth bid_depths = 6 [(gogoproto.customname) = "BidDepths", (gogoproto.jsontag) = "bid_depths"];
  bool is_cover = 7 [(gogoproto.customname) = "IsCover", (gogoproto.jsontag) = "is_cover"];
}

message Depth {
  string price = 1 [(gogoproto.customname) = "Price", (gogoproto.jsontag) = "price"];
  map<string, double> data = 2 [(gogoproto.customname) = "Data", (gogoproto.jsontag) = "data"];
  double volume = 3 [(gogoproto.customname) = "Volume", (gogoproto.jsontag) = "volume"];
}

// 获取单品种行情
message SubscribeOneQuoteReq {
  string exchange = 1 [(gogoproto.customname) = "Exchange", (gogoproto.jsontag) = "exchange"];
  string symbol = 2 [(gogoproto.customname) = "Symbol", (gogoproto.jsontag) = "symbol"];
}

message MultiQuoteData {
  repeated QuoteData quotes = 1 [(gogoproto.customname) = "Quotes", (gogoproto.jsontag) = "quotes"];
}

message QuoteData {
  string symbol = 1 [(gogoproto.customname) = "Symbol", (gogoproto.jsontag) = "symbol"];
  string exchange = 2 [(gogoproto.customname) = "Exchange", (gogoproto.jsontag) = "exchange"];
  int64 msg_seq = 3 [(gogoproto.customname) = "MsgSeq", (gogoproto.jsontag) = "msg_seq"];
  string time = 4 [(gogoproto.customname) = "Time", (gogoproto.jsontag) = "time", (gogoproto.casttype) = "wx/pkg/time.Time"];
  string time_arrive = 5 [(gogoproto.customname) = "TimeArrive", (gogoproto.jsontag) = "time_arrive", (gogoproto.casttype) = "wx/pkg/time.Time"];
  repeated DepthLevel ask_depth = 6 [(gogoproto.customname) = "AskDepth", (gogoproto.jsontag) = "ask_depth"];
  repeated DepthLevel bid_depth = 7 [(gogoproto.customname) = "BidDepth", (gogoproto.jsontag) = "bid_depth"];
  bool is_snap = 8 [(gogoproto.customname) = "IsSnap", (gogoproto.jsontag) = "is_snap"];
}

message DepthLevel {
  Decimal price = 1 [(gogoproto.customname) = "Price", (gogoproto.jsontag) = "price"];
  double volume = 2 [(gogoproto.customname) = "Volume", (gogoproto.jsontag) = "volume"];
}

message Decimal {
  int64 value = 1 [(gogoproto.customname) = "Value", (gogoproto.jsontag) = "value"];
  int32 base = 2 [(gogoproto.customname) = "Base", (gogoproto.jsontag) = "base"];
}